custom/icds_reports/views.py

```diff
@@ -2164,3 +2170,43 @@ def get(self, request, *args, **kwargs):
         except Exception:
             response = dict(isSuccess=False, message='Unknown Error occured')
             return JsonResponse(response, status=500)
+
+
+@location_safe
+class BaseGovernanceAPIView(View):
+    def get_settings(self, request, *args, **kwargs):
+        step = kwargs.get('step')
+        now = datetime.utcnow()
+        month = int(request.GET.get('month', now.month))
+        year = int(request.GET.get('year', now.year))
+
+        if (now.day == 1 or now.day == 2) and now.month == month and now.year == year:
+            prev_month = now - relativedelta(months=1)
+            month = prev_month.month
+            year = prev_month.year
+
+        start = request.GET.get('start', 0)
+
+        return step, start, month, year
+
+
+@method_decorator([api_auth, toggles.mwcd_indicators.required_decorator()], name='dispatch')
+class GovernanceAPIView(BaseGovernanceAPIView):
+
+    def get(self, request, *args, **kwargs):
+        step, start, month, year = self.get_settings(request, *args, **kwargs)
+
+        if step == 'home_visit':
+            length = GOVERNANCE_API_HOME_VISIT_RECORDS_PAGINATION
+            query_filters = {'aggregation_level': 5}
```

Comment:
Actually,
I think also filtering it based on states is what we should do instead. To improve query performance.

Query cost with just state filter reduce significantly.

```
icds_ucr=# EXPLAIN select state_name,district_name,block_name,supervisor_name,awc_name from agg_awc_monthly where month='2019-10-01'  and aggregation_level=5 order by state_name,district_name,block_name,supervisor_name,awc_name limit 20000 OFFSET 10000;
                                                                                                                                                                  QUERY PLAN                                                                                                                                                                   
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1272111.97..1272161.97 rows=20000 width=76)
   ->  Sort  (cost=1272086.97..1282784.03 rows=4278822 width=76)
         Sort Key: awc_location.state_name, awc_location.district_name, awc_location.block_name, awc_location.supervisor_name, awc_location.awc_name
         ->  Merge Left Join  (cost=1.00..466726.63 rows=4278822 width=76)
               Merge Cond: (awc_location.doc_id = agg_awc.awc_id)
               Join Filter: ((months.start_date = agg_awc.month) AND (awc_location.aggregation_level = agg_awc.aggregation_level) AND (awc_location.state_id = agg_awc.state_id) AND (awc_location.district_id = agg_awc.district_id) AND (awc_location.block_id = agg_awc.block_id) AND (awc_location.supervisor_id = agg_awc.supervisor_id))
               ->  Nested Loop  (cost=0.42..160152.85 rows=4278822 width=244)
                     ->  Index Scan using awc_location_local_doc_id_idx on awc_location_local awc_location  (cost=0.42..106642.56 rows=713137 width=240)
                           Filter: (aggregation_level = 5)
                     ->  Materialize  (cost=0.00..25.03 rows=6 width=4)
                           ->  Seq Scan on icds_months_local months  (cost=0.00..25.00 rows=6 width=4)
                                 Filter: (start_date = '2019-10-01'::date)
               ->  Materialize  (cost=0.57..142427.91 rows=742267 width=173)
                     ->  Merge Append  (cost=0.57..140572.25 rows=742267 width=173)
                           Sort Key: agg_awc.awc_id
                           ->  Sort  (cost=0.01..0.02 rows=1 width=168)
                                 Sort Key: agg_awc.awc_id
                                 ->  Seq Scan on agg_awc  (cost=0.00..0.00 rows=1 width=168)
                                       Filter: ((month = '2019-10-01'::date) AND (aggregation_level = 5))
                           ->  Index Scan using "agg_awc_2019-10-01_5_awc_id_idx" on "agg_awc_2019-10-01_5" agg_awc_1  (cost=0.55..133149.55 rows=742266 width=173)
                                 Filter: ((month = '2019-10-01'::date) AND (aggregation_level = 5))
(21 rows)

icds_ucr=# EXPLAIN select state_name,district_name,block_name,supervisor_name,awc_name from agg_awc_monthly where month='2019-10-01'  and aggregation_level=5 and state_id = 'f98e91aa003accb7b849a0f18ebd7039' order by state_name,district_name,block_name,supervisor_name,awc_name limit 20000 OFFSET 10000;
                                                                                                           QUERY PLAN                                                                                                            
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=171987.91..172037.91 rows=20000 width=76)
   ->  Sort  (cost=171962.91..172802.91 rows=336000 width=76)
         Sort Key: awc_location.state_name, awc_location.district_name, awc_location.block_name, awc_location.supervisor_name, awc_location.awc_name
         ->  Merge Left Join  (cost=63094.75..123630.70 rows=336000 width=76)
               Merge Cond: ((awc_location.district_id = agg_awc.district_id) AND (awc_location.block_id = agg_awc.block_id) AND (awc_location.supervisor_id = agg_awc.supervisor_id) AND (awc_location.doc_id = agg_awc.awc_id))
               Join Filter: ((months.start_date = agg_awc.month) AND (awc_location.aggregation_level = agg_awc.aggregation_level) AND (awc_location.state_id = agg_awc.state_id))
               ->  Nested Loop  (cost=0.68..56303.61 rows=336000 width=244)
                     ->  Index Scan using awc_location_local_pkey on awc_location_local awc_location  (cost=0.68..52078.59 rows=56000 width=240)
                           Index Cond: (state_id = 'f98e91aa003accb7b849a0f18ebd7039'::text)
                           Filter: (aggregation_level = 5)
                     ->  Materialize  (cost=0.00..25.03 rows=6 width=4)
                           ->  Seq Scan on icds_months_local months  (cost=0.00..25.00 rows=6 width=4)
                                 Filter: (start_date = '2019-10-01'::date)
               ->  Materialize  (cost=63094.08..63385.08 rows=58200 width=173)
                     ->  Sort  (cost=63094.08..63239.58 rows=58200 width=173)
                           Sort Key: agg_awc.district_id, agg_awc.block_id, agg_awc.supervisor_id, agg_awc.awc_id
                           ->  Append  (cost=0.00..55574.87 rows=58200 width=173)
                                 ->  Seq Scan on agg_awc  (cost=0.00..0.00 rows=1 width=168)
                                       Filter: ((month = '2019-10-01'::date) AND (aggregation_level = 5) AND (state_id = 'f98e91aa003accb7b849a0f18ebd7039'::text))
                                 ->  Index Scan using "agg_awc_2019-10-01_5_state_id_district_id_block_id_supervis_idx" on "agg_awc_2019-10-01_5" agg_awc_1  (cost=0.68..55283.87 rows=58199 width=173)
                                       Index Cond: (state_id = 'f98e91aa003accb7b849a0f18ebd7039'::text)
                                       Filter: ((month = '2019-10-01'::date) AND (aggregation_level = 5))
(22 rows)


```

Reply:
could. you take state_site_code(state_site_code because state_names are alpha strings and numbers are reliable for passing as param) as a query param through URL and find the state_id through it and then query the view with state_id.

Reply:
we would also have to create a url to get all the latest states info(if there already not a view providing that info).

Reply:
> could. you take state_site_code(state_site_code because state_names are alpha strings and numbers are reliable for passing as param) as a query param through URL and find the state_id through it and then query the view with state_id.

done

Reply:
> we would also have to create a url to get all the latest states info(if there already not a view providing that info).

what is the expected json format needed?

Reply:
I think we can do it in separate PR.